src/VisualStudio/Core/Def/ProjectSystem/OpenTextBufferProvider.cs

```diff
@@ -0,0 +1,313 @@
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    /// <summary>
+    /// A class that provides access to the currently open list of files in Visual Studio. You are able to ask for the text buffer for a document on any thread; events are raised
+    /// on thread, and any method that provides a <see cref="IVsHierarchy"/> must be used on the UI thread.
+    /// </summary>
+    [Export(typeof(OpenTextBufferProvider))]
+    internal sealed class OpenTextBufferProvider : IVsRunningDocTableEvents3, IDisposable
+    {
+        private bool _isDisposed = false;
+
+        /// <summary>
+        /// A simple object for asserting when we're on the UI thread.
+        /// </summary>
+        private readonly ForegroundThreadAffinitizedObject _foregroundAffinitization;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IVsRunningDocumentTable4 _runningDocumentTable;
+
+        private ImmutableArray<IOpenTextBufferEventListener> _listeners = ImmutableArray<IOpenTextBufferEventListener>.Empty;
+
+        /// <summary>
+        /// The map from monikers to open text buffers; because we can only fetch the text buffer on the UI thread, all updates to this must be done from the UI thread.
+        /// </summary>
+        private ImmutableDictionary<string, ITextBuffer> _monikerToTextBufferMap = ImmutableDictionary<string, ITextBuffer>.Empty.WithComparers(StringComparer.OrdinalIgnoreCase);
+
+        private uint _runningDocumentTableEventsCookie;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public OpenTextBufferProvider(
+            IThreadingContext threadingContext,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _foregroundAffinitization = new ForegroundThreadAffinitizedObject(threadingContext, assertIsForeground: false);
+
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+
+            // The running document table since 16.0 has limited operations that can be done in a free threaded manner, specifically fetching the service and advising events.
+            // This is specifically guaranteed by the shell that those limited operations are safe and do not cause RPCs, and it's important we don't try to fetch the service
+            // via a helper that will "helpfully" try to jump to the UI thread.
+            var runningDocumentTable = (IVsRunningDocumentTable)serviceProvider.GetService(typeof(SVsRunningDocumentTable));
```

Comment:
This is the actual IServiceProvider.GetService() that appears to handle the background-thread case; I'm confirming with @ryanmolden.

Reply:
Confirmed with some help setting well-placed breakpoints.